pennylane_lightning/core/src/utils/NDPermuter.hpp

```diff
@@ -0,0 +1,268 @@
+        return permuter_b_.Transpose(data_in, shape, current_order, new_order);
+    }
+
+  protected:
+    friend PermuterBackend;
+
+  private:
+    PermuterBackend permuter_b_;
+};
+
+/**
+ * @brief Default Permuter backend class for generalised transforms. Adapted
+ * from QFlex.
+ *
+ * @tparam blocksize Controls the internal data chunk size for cache blocking.
+ */
+template <size_t BLOCKSIZE = 1024> class DefaultPermuter {
+
+  public:
+    /**
+     * @brief Reference-based transpose operation. See `Permuter` class for more
+     * details.
+     */
+    template <class T>
+    void Transpose(const std::vector<T> &data_,
+                   const std::vector<size_t> &shape, std::vector<T> &data_out,
+                   const std::vector<std::string> &old_indices,
+                   const std::vector<std::string> &new_indices) {
+        data_out = data_;
+
+        if (new_indices == old_indices)
+            return;
+
+        const std::size_t num_indices = old_indices.size();
+        const std::size_t total_dim = data_.size();
+        std::size_t remaining_data = total_dim;
+
+        if (num_indices == 0) {
+            PL_ABORT("Number of indices cannot be zero.");
+        }
+
+        // Create map_old_to_new_idxpos from old to new indices, and
+        // new_dimensions.
+        std::vector<size_t> map_old_to_new_idxpos(num_indices);
+        std::vector<size_t> new_dimensions(num_indices);
+        for (size_t i = 0; i < num_indices; ++i) {
+            for (size_t j = 0; j < num_indices; ++j) {
+                if (old_indices[i] == new_indices[j]) {
+                    map_old_to_new_idxpos[i] = j;
+                    new_dimensions[j] = shape[i];
+                    break;
+                }
+            }
+        }
+
+        std::vector<size_t> old_super_dimensions(num_indices, 1);
+        std::vector<size_t> new_super_dimensions(num_indices, 1);
+
+        const std::size_t old_dimensions_size = shape.size();
+        for (size_t i = old_dimensions_size; --i;) {
+            old_super_dimensions[i - 1] = old_super_dimensions[i] * shape[i];
+            new_super_dimensions[i - 1] =
+                new_super_dimensions[i] * new_dimensions[i];
+        }
+
+        std::vector<std::size_t> small_map_old_to_new_position(blocksize_);
+
+        // Position old and new.
+        std::size_t po = 0, pn;
+        // Counter of the values of each indices in the iteration (old
+        // ordering).
+        std::vector<size_t> old_counter(num_indices, 0);
+        // offset is important when doing this in blocks, as it's indeed
+        // implemented.
+        std::size_t offset = 0;
+        // internal_po keeps track of interations within a block.
+        // Blocks have size `blocksize`.
+        std::size_t internal_po = 0;
+
+        T *data = data_out.data();
+        const T *scratch =
+            data_.data(); // internal pointer offers better performance than
```

Comment:
```suggestion
            data_in.data(); // internal pointer offers better performance than
```

Reply:
To be resolved by later importing directly from JET. Resolving.